### PR TITLE
Fix Gemma-4-26B-A4B full-model PCC regression — revert #49715 (LLK fast reduce unpack)

### LINKED ISSUE #50255: Gemma-4-26B-A4B full-model PCC regression on scheduled T2 E2E — interim revert of #49715
## Regression

**Gemma-4-26B-A4B** `[wh_llmbox_perf]` `test_full_model` on the (Tier 2) Models E2E scheduled pipeline: full-model PCC dropped from a bit-stable **0.9383** (7 consecutive runs) to **0.7696** (threshold 0.8), first failing 2026-07-16.

## Cause

Bisected to **#49715** ("[LLK] fast reduce unpack for 32x32 tiles", `fe7519882b5`). Its 4-face `REDUCE_ROW` SUM/AVG fast path never seeds the SrcB CR-base, so the right-column faces (F1/F3) are not reduced. Full root-cause analysis + a failed surgical-fix attempt are in **#50224**.

## Resolution

**PR #50085** — clean revert of #49715 — restores PCC to 0.9383 (CI-validated). This issue tracks that interim mitigation.

- Interim fix (this issue): revert → #50085
- Proper fast-path fix (re-land the optimization correctly): tracked in **#50224**, owned by @handrewsTT / @markoradosavljevic

Owner: @handrewsTT (gemma4)

### PR DESCRIPTION
**Pull request by mtairum**
## Bad commit

🔴 **#49715 — "[LLK] fast reduce unpack for 32x32 tiles"** (`fe7519882b53d7d42342f71da42e9850a1c6e2e4`, @markoradosavljevic, 2026-07-15)

Identified via CI bisect on the (Tier 2) Models E2E scheduled pipeline (window `52c1d0e554..5f990184fc`): p35 GOOD (PCC 0.9383 @ `ed4331a`) → p36 BAD (PCC 0.7696 @ `fe75198`); #49715 is the lone commit between them.

## Regression

**Gemma-4-26B-A4B** `[wh_llmbox_perf]` `test_full_model`: full-model PCC dropped from a bit-stable **0.9383** (7 consecutive scheduled runs) to **0.7696** (threshold 0.8), first failing 2026-07-16. Prefill/decode sub-checks (0.99) still pass → break is in the full-model logits path.

## Cause

#49715's 4-face `REDUCE_ROW` SUM/AVG fast path never seeds the SrcB CR-base to row 16, so its `cr` reload re-reads faces F0/F2 and the right-column faces F1/F3 are never reduced. Full root-cause analysis + a failed surgical-fix attempt: **#50224**.

## This PR

Clean **revert of #49715** — restores the correct general reduce path (known-good bit-stable 0.9383). No LLK signature changes → metal callers unaffected. The perf optimization should be re-landed with corrected fast-path microcode (**#50224**, owner @handrewsTT / @markoradosavljevic).

**CI validation:** run [29582263968](https://github.com/tenstorrent/tt-metal/actions/runs/29582263968) → full-model PCC back to **0.9383** ✅

Closes #50255
Root cause / proper fast-path fix tracked in #50224

### CODE REVIEW

**Comment by mtairum** [comment]
### Bisect update 1

**p24** (`0e89eb3c2e8`, #49731, pos 24/48) → full-model PCC **0.9383 = GOOD** (pre-regression).

Culprit localized to the **newer half, pos 25-48** (p24 → tip `5f990184fc`). `p12` (pos 12) is now redundant (older than confirmed-good p24).

Decisive next signal = **p36** (`fe7519882b5`, #49715 LLK reduce, pos 36):
- p36 **BAD** → culprit in pos 25-36 (incl. #48930 untilize padded-tensor fix)
- p36 **GOOD** → culprit in pos 37-48 (incl. #49914 SFPI 7.67.0, #49635 LLK tilize)

_(Note: p24's first attempt died on a runner comms-loss build-infra failure; this is the clean re-run.)_

**Comment by mtairum** [comment]
### Bisect update 2 — culprit range halved

Round-1 results:
| point | commit | PR | PCC | verdict |
|---|---|---|---|---|
| p12 | `ee4e290` | #49750 | 0.9383 | GOOD |
| p24 | `0e89eb3` | #49731 | 0.9383 | GOOD |
| p36 | `fe75198` | #49715 | **0.7696** | **BAD** |

→ **Culprit ∈ pos 25-36** (12 commits). Prime numerical suspects in range: **#48930** untilize-padded (pos34), **#49715** LLK fast-reduce (pos36), **#49876** all_to_all_combine (pos33, relevant — A4B is MoE).

Round-2 (2 parallel narrowing runs):
| point | commit | PR | run | result |
|---|---|---|---|---|
| p33 | `5e05020e15a` | #49876 | https://github.com/tenstorrent/tt-metal/actions/runs/29514491513 | pending |
| p35 | `ed4331a0166` | #49730 | https://github.com/tenstorrent/tt-metal/actions/runs/29514494310 | pending |

Decode: p35 GOOD → **#49715**; p35 BAD & p33 GOOD → **#48930**; p33 BAD → range pos25-33.

**Comment by mtairum** [comment]
### ✅ Culprit pinned: #49715

Round-2 results:
| point | commit | PR | PCC | verdict |
|---|---|---|---|---|
| p33 | `5e05020` | #49876 | — | NO-SIGNAL (runner comms-loss build infra flake) |
| p35 | `ed4331a` | #49730 | 0.9383 | GOOD |
| p36 | `fe75198` | #49715 | 0.7696 | BAD |

p35 GOOD + p36 BAD ⇒ the **single commit between them is the culprit**:

**#49715 — "[LLK] fast reduce unpack for 32x32 tiles"** (`fe7519882b53d7d42342f71da42e9850a1c6e2e4`, Marko Radosavljevic, 2026-07-15).

Changed files (reduce LLK, both archs):
- `tt_metal/tt-llk/.../tt_llk_wormhole_b0/llk_lib/llk_math_reduce.h`
- `tt_metal/tt-llk/.../tt_llk_wormhole_b0/llk_lib/llk_unpack_AB_reduce.h`
- (+ blackhole equivalents, + reduce_perf.cpp)

Mechanism matches: the reduce LLK feeds softmax / normalization, so a reduce-unpack numerical change shifts the full-model logits — dropping `test_full_model` PCC 0.9383 → 0.7696 (threshold 0.8) while the higher-threshold prefill/decode sub-checks (0.99) already tripped were... n/a. Gemma-4-26B-A4B runs on WH, so the `tt_llk_wormhole_b0` reduce edit is the operative one.

**Scope note:** #49715 touches the shared reduce LLK on both WH and BH — other reduce-heavy models may be affected; worth a broader accuracy check.

**Recommended action:** flag Marko Radosavljevic; either revert #49715 or land an LLK reduce-numerics fix. Bit-exact stable 0.9383 for 7 prior days argues the pre-#49715 behavior was correct.

**Comment by mtairum** [comment]
### Fix pushed — revert of #49715

Branch reset onto current `origin/main` (which contains the bug) and applied a clean **revert of #49715** (`fe7519882b5`). Restores the correct general reduce path for `REDUCE_ROW` SUM/AVG; removes the broken 4-face fast path on WH + BH.

Why revert vs a surgical disable: this is delicate reduce microcode where the math and unpack fast paths must stay in lockstep, and an `if(false)`-style disable risks `-Werror` unreachable-code failures and leaves the possibility that the rewritten general path itself diverges. A revert restores the exact numerics that were bit-stable (0.9383) for the 7 prior scheduled runs. The perf optimization should be re-landed once the fast-path microcode is corrected.

Files reverted (5): WH+BH `llk_math_reduce.h`, WH+BH `llk_unpack_AB_reduce.h`, `reduce_perf.cpp` (test expectations). No LLK function signatures changed → metal callers unaffected.

Tracking issue: #50224

**CI validation** (expect full-model PCC back to 0.9383): https://github.com/tenstorrent/tt-metal/actions/runs/29582263968

**Comment by mtairum** [comment]
### ✅ CI validation PASSED

Run [29582263968](https://github.com/tenstorrent/tt-metal/actions/runs/29582263968) — `gemma-4-26b-a4b [wh_llmbox_perf]`, conclusion **success**:

```
test_full_model:444 - Full model PCC (seq_len=5): 0.9382964432181848
```

Full-model PCC restored to the pre-#49715 value **0.9383** (bit-identical to the 7 good pre-regression runs), passing the 0.8 gate. Revert confirmed as the fix.

Kept as **draft** pending review. Ready to mark for review whenever you'd like; #49715 author (@markoradosavljevic) should re-land the reduce fast-path optimization with corrected microcode. Closes #50224 on merge.
